### PR TITLE
Resolve all native promise to avoid memory leak

### PR DESCRIPTION
**Pull request by birdy-**

### CODE REVIEW

File: CancelablePromise.js [hunk 1]
```diff
@@ -19,19 +19,36 @@ export default class CancelablePromise {
   }
 
   then(success, error) {
-    return new CancelablePromise((resolve, reject) => {
+    const p = new CancelablePromise((resolve, reject) => {
       this._promise.then((r) => {
+        if (this._canceled) {
+          p.cancel();
+          resolve(r);
```
Comment (bobinette) [inline CancelablePromise.js]: why forward `r`?

File: CancelablePromise.js [hunk 2]
```diff
@@ -19,19 +19,36 @@ export default class CancelablePromise {
...
+        }
         if (success && !this._canceled) {
+          let argument = r;
           const returned = success(r);
-          resolve(returned || r);
+          if (typeof returned !== 'undefined') {
+            argument = returned;
+          }
+          resolve(argument);
```
Comment (bobinette) [inline CancelablePromise.js]: if returned is `undefined`, we should call `resolve()`, not forward the current argumet. What do you think?

File: CancelablePromise.js [hunk 3]
```diff
@@ -19,19 +19,36 @@ export default class CancelablePromise {
...
+          reject(r);
+        }
         if (error && !this._canceled) {
+          let argument = r;
           const returned = error(r);
-          reject(returned || r);
+          if (typeof returned !== 'undefined') {
+            argument = returned;
+          }
+          reject(argument);
```
Comment (bobinette) [inline CancelablePromise.js]: same

File: CancelablePromise.spec.js [hunk 4]
```diff
@@ -106,10 +106,16 @@ describe(__filename, () => {
     rejectedPromise.cancel();
     rejectedPromise.then(() => {
       hasFailed = true;
-      done(new Error('Callback should not be executed'));
+      done(new Error('Success callback should not be executed'));
```
Comment (bobinette) [inline CancelablePromise.spec.js]: what happens in the other way around:
```javascript
p.then(() => {done()})
p.cancel()
```

maybe with a `setTimeout` to ensure `cancel` is called before the execution of the `then` :wink:

File: CancelablePromise.js (hunk 1 above)
Comment (birdy-) [inline CancelablePromise.js]: why not ?

File: CancelablePromise.js (hunk 2 above)
Comment (sebastiendavid) [inline CancelablePromise.js]: this is the same

File: CancelablePromise.spec.js (hunk 4 above)
Comment (birdy-) [inline CancelablePromise.spec.js]: cancel is synchronous

**Pull request merged**